rinad/src/ipcm/ipcm.cc

```diff
@@ -1769,4 +1786,266 @@ void IPCManager_::io_loop(){
 
 }
 
+CatalogPsInfo::CatalogPsInfo(const rina::PsInfo& psinfo,
+			     map<string, CatalogPlugin>::iterator plit)
+				: PsInfo(psinfo)
+{
+	plugin = plit;
+}
+
+static bool endswith(const string& s, const string& suffix)
+{
+	if (s.size() < suffix.size()) {
+		return false;
+	}
+
+	return suffix == s.substr(s.size() - suffix.size(), suffix.size());
+}
+
+#define MANIFEST_SUFFIX ".manifest"
+
+void Catalog::import()
+{
+	const rinad::RINAConfiguration& config = IPCManager->getConfig();
+
+	for (list<string>::const_iterator lit =
+		config.local.pluginsPaths.begin();
+			lit != config.local.pluginsPaths.end(); lit++) {
+		DIR *dir = opendir(lit->c_str());
+		struct dirent *ent;
+
+		if (dir == NULL) {
+			LOG_WARN("Failed to open plugins directory: '%s'",
+				 lit->c_str());
+			continue;
+		}
+
+		while ((ent = readdir(dir)) != NULL) {
+			string plugin_name(ent->d_name);
+
+			if (!endswith(plugin_name, MANIFEST_SUFFIX)) {
+				continue;
+			}
+
+			LOG_INFO("Catalog: found manifest %s",
+					plugin_name.c_str());
+
+			// Remove the MANIFEST_SUFFIX suffix
+			plugin_name = plugin_name.substr(0,
+						plugin_name.size() -
+						string(MANIFEST_SUFFIX).size());
+
+			add_plugin(plugin_name, *lit);
+		}
+
+		closedir(dir);
+	}
+}
+
+void Catalog::add_plugin(const string& plugin_name, const string& plugin_path)
+{
+	map<string, CatalogPlugin>::iterator plit;
+	list<rina::PsInfo> new_policy_sets;
+	rina::WriteScopedLock wlock(rwlock);
+	int ret;
+
+	ret = rina::plugin_get_info(plugin_name, plugin_path, new_policy_sets);
+	if (ret) {
+		LOG_WARN("Failed to load manifest file for plugin '%s'",
+			 plugin_name.c_str());
+		return;
+	}
+
+	if (plugins.count(plugin_name)) {
+		// Plugin already in the catalog
+		return;
```

Comment:
Regarding the get_info call, yes good catch --> changing it
Regarding the plugin_get_info, I would keep the "return -1" model, it's ok for our purposes.